Add `--toolchain` option

tkrajacic opened the pull request:
Adds the ability to specify a custom Swift toolchain via a `--toolchain` option on the `build`, `update`, and `archive` commands.

E.g.: `carthage build --toolchain org.swift.3020160509a`

**Todo**
- [x] Make toolchain parameter default to `nil` where desired
- [x] tests?


**Resolves issue** #1039

tkrajacic commented:
Since I haven't updated the tests, and the function arguments require the `toolchain` parameter at the moment, ci fails of course.
I'll update that after initial feedback if the general approach is fine.

mdiep commented on Source/CarthageKit/Xcode.swift:
```diff
@@ -171,14 +171,14 @@ public func schemesInProject(project: ProjectLocator) -> SignalProducer<String,
 			return schemesInProject(project)
 				.flatMap(.Merge) { scheme -> SignalProducer<String, CarthageError> in
-					let buildArguments = BuildArguments(project: project, scheme: scheme, configuration: configuration)
+					let buildArguments = BuildArguments(project: project, scheme: scheme, configuration: configuration, toolchain: toolchain)
```
Is this actually needed? It doesn't seem like the toolchain would affect whether we'd build the scheme.

mdiep commented on Source/carthage/Archive.swift:
```diff
@@ -14,19 +14,21 @@ import ReactiveCocoa
 
 		public static func evaluate(m: CommandMode) -> Result<Options, CommandantError<CarthageError>> {
 			return create
+				<*> m <| Option<String?>(key: "toolchain", defaultValue: nil, usage: "the toolchain to build with")
```
`carthage archive` doesn't build anything, so it shouldn't need the `toolchain` option.

mdiep commented:
This is a great first step! ☺️

tkrajacic commented:
I removed the unnecessary `toolchain` parameters. They were remnants from when I was hunting down compile errors ;)
I also added a default `nil` to the `BuildArgument` constructor and removed all explicit `nil` passing.

tkrajacic commented:
I didn't add the default `nil` to `buildDependencyProject(...)` and `buildInDirectory(...)` because _CarthageKit_ always passes in the toolchain from a higher level command.

When adding an expectation to the _BuildArgumentsSpec_ though I saw that the test don't do that and call those methods directly.

Does it make sense to add the default `nil` value to the `toolchain` parameter as well on `buildDependencyProject(...)` and `buildInDirectory(...)`? In my mind it doesn't make sense to leave out the toolchain anywhere downstream because once it has been supplied as an argument it must be used everywhere (all build products need to be compiled with the same toolchain).

mdiep commented:
Let's add `nil` defaults to all the `public` functions that take a toolchain. That way we don't break compatibility for CarthageKit.

Also, I missed this yesterday, but `carthage bootstrap` needs a `--toolchain` argument as well.

This is shaping up great!

tkrajacic commented:
Doesn't `bootstrap` already work with all of `update`s commands?
It reuses the `UpdateCommand.Options`, and running `bootstrap --toolchain xxx` did work when I tried it.

mdiep commented:
> Doesn't bootstrap already work with all of updates commands?
> It reuses the UpdateCommand.Options, and running bootstrap --toolchain xxx did work when I tried it.

Oh, duh!

mdiep commented:
This is great! Thanks so much for the pull request! ✨

NachoSoto commented:
What's the expected value for this parameter? I'm trying "swift-DEVELOPMENT-SNAPSHOT-2016-09-06-a", as well as the full path, but Carthage is ignoring it and using the default toolchain.

tkrajacic commented:
For the tests I had `org.swift.3020160509a` but I don't immediately recall how I found this. (I am already in bed over here so I can't check right now)

I'll see what I can find in the morning.

ikesyo commented:
> What's the expected value for this parameter?

The `CFBundleIdentifier` value of `Info.plist` located at `/Library/Developer/Toolchains/swift-DEVELOPMENT-SNAPSHOT-2016-MM-dd-a.xctoolchain/`.
